src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnectionOptions.cs

```diff
@@ -131,6 +132,17 @@ internal override void Validate(string argumentName)
         {
             throw new ArgumentNullException(SR.Format(SR.net_quic_not_null_open_connection, nameof(QuicClientConnectionOptions.RemoteEndPoint)), argumentName);
         }
+
+        // MsQuic is using DualMode sockets and that will faill even for IPv4 if AF_INET6 is not available.
```

Comment:
If quic can't work without OSSupportsIPv6, shouldn't that instead be part of QuicConnection.IsSupported, e.g.
```C#
public static bool IsSupported => MsQuicApi.IsQuicSupported && Socket.OSSupportsIPv6;
```

And for RemoteEndPoint, shouldn't that validation be done as part of setting the RemoteEndPoint, e.g. move the validation of its AddressFamily into the property setter?

Reply:
I can update `IsSupported`. I realized that only after writing tests and testing various setups. The only caveat is that when 'IsSupported` is `false` it is very difficult IMHO to figure out why. But I guess we can document it in remarks.

As far as validation in setter: AFAIK we don't do that or anything else - neither in Quic nor SslStream e.g. for property bags we validate when applied to some action now when they are created. We _could_ go down that route but then I would expect same for other properties.

Reply:
Actually, I think moving `Socket.OSSupportsIPv6` to `IsSupported` makes the most sense. AFAICT, it cannot change during app run and is essential for MsQuic to work. 
But I'd suggest to make it part of `MsQuicApi.IsQuicSupported` as this might be MsQuic specific behavior and it will also give us opportunity to log this as we do with missing libmsquic etc. This should also answer the problem of diagnosing this as it would be evident from the logs what's going on. WDYT?

Reply:
make sense. How about the IPv4 check? It seems like less likely but sill possible. And than the question if it is ok to throw synchronously, store exception or do it when somebody is manipulating the option object.

Reply:
I would not put it into the setter, we intentionally postpone all validations until the options are used.

We do other IP related and other checks afterwards, e.g.:
https://github.com/dotnet/runtime/blob/7c00b17be1b2ffb6ed49ad68cf36e9a056323152/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs#L243-L247
https://github.com/dotnet/runtime/blob/7c00b17be1b2ffb6ed49ad68cf36e9a056323152/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs#L263-L266

I still personally think this fits there better.

Reply:
I moved the IPv6 check to IsSupported as suggested. As fas as the other check I'm still not convinced. The link you posted spawns task to figure out if something is going to work. But for the address family we can tell upfront it is going to fail (without any additional work) 

From my prospective if something is going to fail the sooner we surface error and we stop doing extra work the better. Throwing in validation avoids creation of the task and and executing more code. But I can move it if that is consensus.

Reply:
I left IPv6 check as part of the IsSupported and I took out IPv4 check. I'll make plan and I'll put it up together with fixes for https://github.com/dotnet/runtime/issues/74053.